python/cudf/cudf/utils/utils.py

```diff
@@ -490,3 +490,22 @@ def _create_pandas_series(
         copy=copy,
         fastpath=fastpath,
     )
+
+
+def _maybe_indices_to_slice(indices: cp.ndarray) -> Union[slice, cp.ndarray]:
+    """Makes best effort to convert an array of indices into a python slice.
+    If the conversion is not possible, return input. `indices` are expected
+    to be valid.
+    """
+    # TODO: improve efficiency by avoiding sync.
+    if len(indices) == 1:
+        x = indices[0].item()
+        return slice(x, x + 1)
+    if len(indices) == 2:
+        x1, x2 = indices[0].item(), indices[1].item()
+        return slice(x1, x2 + 1, x2 - x1)
+    start, step = indices[0].item(), (indices[1] - indices[0]).item()
+    stop = start + step * len(indices)
+    if (indices == cp.arange(start, stop, step)).all():
```

Comment:
Should we use numpy here instead of cupy? Just thinking about our previous discussion on syncing.

Reply:
Here's what I found after running some benchmarks. On a very high level, `_maybe_indices_to_slice` only make up for a small fraction of the total cost of `get_loc`. The majority of the time is spent on sorting and and gathering. So it overall it does not make a big difference optimizing for this section. (see red circle area, benchmarked at 100M rows)

![image](https://user-images.githubusercontent.com/13521008/121960960-a6222a00-cd1b-11eb-974d-3e6fcc9add0e.png)

But if we are to look at the details, we can compare just the post-processing part. That is, the time spent on judging if the indices can be converted into a slice, if not, create a boolean mask. And here are the results:
```
GPU:
100000
slice/boolean mask post processing 0.0046520233154296875
1000000
slice/boolean mask post processing 0.001522064208984375
10000000
slice/boolean mask post processing 0.004399299621582031
100000000
slice/boolean mask post processing 0.018871545791625977

CPU:
100000
slice/boolean mask post processing 0.0006105899810791016
1000000
slice/boolean mask post processing 0.0025777816772460938
10000000
slice/boolean mask post processing 0.020697832107543945
100000000
slice/boolean mask post processing 0.21960783004760742
```
And we can see using cupy here has some advantage when array is large. Plus we might further improve the performance if we can figure out how to call `cupy.arange` without device syncs.

Benchmark details:
the CPU path of the `_maybe_indices_to_slice` is:
```python
def _maybe_indices_to_slice(indices: cp.ndarray) -> Union[slice, cp.ndarray]:
    # cpu path:
    indices = indices.get()
    if len(indices) == 1:
        x = indices[0]
        return slice(x, x + 1)
    if len(indices) == 2:
        x1, x2 = indices[0], indices[1]
        return slice(x1, x2 + 1, x2 - x1)
    start, step = indices[0], (indices[1] - indices[0])
    stop = start + step * len(indices)
    if (indices == np.arange(start, stop, step)).all():
        return slice(start, stop, step)
    return indices
```
And later use `np.full` instead of `cp.cull`.

Benchmark code:
```python
import cupy as cp
import pandas as pd
import cudf

def make_non_monotonic_non_unique_multiindex(N):
    half_n = int(N / 2)
    l0 = cp.concatenate([cp.arange(0, half_n), cp.full(half_n, 42)])
    l1 = cp.concatenate([cp.arange(half_n, N), cp.full(half_n, 42)])
    random_order = cp.arange(0, N)
    cp.random.shuffle(random_order)
    l0[random_order] = l0
    l1[random_order] = l1
    l0_arr = l0.get()
    l1_arr = l1.get()
    return pd.MultiIndex.from_arrays([l0_arr, l1_arr])

cp.random.seed(0) # reproducibility
target = (42, 42)
for N in [10**5, 10**6, 10**7, 10**8]:
    print (N)
    pidx = make_non_monotonic_non_unique_multiindex(N)
    gidx = cudf.from_pandas(pidx)
    gidx.get_loc(target)
    del pidx, gidx
```

Reply:
Sounds good! I guess in this case the array could be quite large so that makes sense.

Reply:
Would you like to figure out the efficiency issues with MultiIndex in a separate PR? I think for regular indices this looks good to me.

Reply:
> Would you like to figure out the efficiency issues with MultiIndex in a separate PR?

Yup, that sounds good.

Reply:
Ok, great!